### DISCUSSION TITLE
vlangroup in api

### DISCUSSION THREAD

**Discussion by jskoetsier** [netbox-community/netbox#7518] [Q&A]
First of all, forgive my bad coding, but i'm trying to migrate prefixes from an old netbox environment to a new netbox environement with the pynetbox api, feeling a bit lost on how to get the vlan into the right vlan group, does anyone have some pointers on this? I have multiple VRFs, VLAN Groups with overlapping VLAN's.

My code below:
```
import pynetbox

old_netbox='http://10.x/'
new_netbox='http://127.x/'
onb = pynetbox.api(old_netbox, token="*",threading=True)
nnb = pynetbox.api(new_netbox, token="*",threading=True)
all_prefixes = onb.ipam.prefixes.all()

for prefix in all_prefixes:
    purple = 3
    dcs = 4
    transit = 6
    if prefix.vlan is None:
        prefix_vlan = 1
    else:
        prefix_vlan = prefix.vlan.vid
    if prefix.vrf is None:
        prefix_vrf = 1
    else:
        prefix_vrf = prefix.vrf.id
    if prefix_vrf is purple:
        prefix_vrfid = 1
    elif prefix_vrf is dcs:
        prefix_vrfid = 2
    elif prefix_vrf is transit:
        prefix_vrfid = 5
    else:
        prefix_vrfid = 1
    try:
        prefixcreate = nnb.ipam.prefixes.create({"prefix": prefix.prefix, "vrf": prefix_vrfid, "vlan": prefix_vlan})
    except pynetbox.RequestError as e:
        print ("Could not create prefix, error: {}".format(str(e)))
```

**Comment by candlerb** [comment]
Sidestepping your question: do you want to migrate the *entire* contents of the old netbox environment? If so it would be far quicker and easier just to [replicate the database](https://netbox.readthedocs.io/en/stable/administration/replicating-netbox/) and then run `./upgrade.sh` to update it to match the new version.  If going from pre-2.11 to 3.0, you need to go via 2.11.

Regarding the specific question: firstly, please can you edit your post, and add three backticks (` ``` `) on a line of their own, before and after your code.  That will fix the formatting, which is obviously critical for python code.

Your code seems to show you wanting to associate a prefix with a vlan (in which case, you'll need to create the vlan object in your database first).  But your question asked about associating vlans with vlangroups.  So I'm afraid the two don't join up for me.

Also with vlans, you need to be very careful of the distinction between `vid` (which is the 802.11q tag of the VLAN), and the vlan `id` which is the unique identifier of the vlan object in Netbox.  You can have multiple VLANs with the same `vid` - for example, you might have two separate vlans in two different sites, but they both have vid 200.  These would be two different objects in Netbox, with their own database `id`.

**Reply by jskoetsier** [reply]
Hi, 

Fixed the post ;)

The VLANs already exist in the new IPAM (as well as the VLAN groups, VRF's etc). So, thinking about it a bit (because, unfortunately it's not 1 netbox instance i'm migrating from, I have to migrate 3 of them and some php-ipam), I think that I will add an extra lookup of a VLAN in a specific group and try to correlate the old vid with the new id and then create the prefix with the new id instead of the vid.

**Reply by candlerb** [reply]
Yes, that sounds right.  I can see you're mapping VRF IDs from those in the old netbox to the new one; a corresponding process will be needed for VLANs.

Note that  "related objects may be specified by either numeric ID (primary key), or by a set of attributes sufficiently unique to return the desired object." - see [REST API docs](https://netbox.readthedocs.io/en/stable/rest-api/overview/#related-objects)

So in principle you ought to be able to write something like this, if pynetbox allows it:

```
nnb.ipam.prefixes.create({"prefix": prefix.prefix, "vrf": prefix_vrfid, "vlan": {"vid": x, "vlangroup": y}})
```

I don't know if that works down multiple levels though:

```
nnb.ipam.prefixes.create({"prefix": prefix.prefix, "vrf": prefix_vrfid, "vlan": {"vid": x, "vlangroup": {"name": y}}})
```

**Reply by jskoetsier** [reply]
@jeremystretch wouldn't it be much easier to have vlan_group in the ipam/prefixes/vlan when the vlan exists if you have multiple vlan groups and you want to sync stuff between those (because ID's differ)
